listenbrainz/webserver/static/js/src/UserArtists.tsx

```diff
@@ -0,0 +1,153 @@
+  private maxListens = 0; // Number of listens for first artist used to scale the graph
+  private totalPages = 0; // Total namber of pages, useful for pagination
+  private currPage = 1; // Current page, 1 by default
+
+  constructor(props: UserArtistsProps) {
+    super(props);
+
+    this.APIService = new APIService(
+      props.apiUrl || `${window.location.origin}/1`
+    ); // Used to access LB API
+
+    this.baseUrl = window.location.origin + window.location.pathname;
+
+    this.state = {
+      data: [],
+    };
+  }
+
+  async componentDidMount() {
+    const { user } = this.props;
+
+    // Fetch page number from URL
+    const url = new URL(window.location.href);
+    if (url.searchParams.get("page")) {
+      this.currPage = Number(url.searchParams.get("page"));
+    } else {
+      this.currPage = 1;
+    }
+
+    // Fetch data from backend
+    const offset = (this.currPage - 1) * this.ROWS_PER_PAGE;
+    try {
+      let data = await this.APIService.getUserStats(
+        user.name,
+        undefined,
+        undefined,
+        1
+      );
+      this.maxListens = data.payload.artists[0].listen_count;
+      this.totalPages = Math.ceil(
+        data.payload.total_artist_count / this.ROWS_PER_PAGE
+      );
+      data = await this.APIService.getUserStats(
+        user.name,
+        undefined,
+        offset,
+        this.ROWS_PER_PAGE
+      );
+      this.setState({ data: this.processData(data, offset) });
+    } catch (error) {
+      // Error Boundaries don't catch errors in async code.
+      // Throwing an error in setState fixes this.
+      // This is a hacky solution but should be fixed with upcoming concurrent mode in React.
+      this.setState(() => {
+        throw error;
+      });
+    }
+  }
+
+  processData = (
+    data: UserArtistsResponse,
+    offset: number
+  ): UserArtistsData => {
+    const result = data.payload.artists
+      .map((elem, idx: number) => {
+        return {
+          id: `${offset + idx + 1}. ${elem.artist_name}`,
+          count: elem.listen_count,
+        };
+      })
+      .reverse();
+    return result;
+  };
+
+  render() {
+    const { data } = this.state;
+    const prevPage = this.currPage - 1;
+    const nextPage = this.currPage + 1;
+
+    return (
+      <div>
+        <div className="row">
+          <div className="col-md-8" style={{ height: "75em" }}>
+            <Bar data={data} maxValue={this.maxListens} />
+          </div>
+        </div>
+        <div className="row">
+          <div className="col-md-8">
+            <ul className="pager">
+              <li className={`previous ${!(prevPage > 0) ? "hidden" : ""}`}>
+                <a href={`${this.baseUrl}?page=${prevPage}`}>&larr; Previous</a>
+              </li>
+              <li
+                className={`next ${
+                  !(nextPage <= this.totalPages) ? "hidden" : ""
+                }`}
+              >
+                <a href={`${this.baseUrl}?page=${nextPage}`}>Next &rarr;</a>
```

Comment:
Have you thought about how you could load the next/previous page without reloading the page?
I think if you extract some of the mechanisms in `componentDidMount` to be reusable (specifically the data loading), you could simply call it again with a page offset and that would recreate the graph (considering `this.state.data` would change).

That might also reduce the number of calls to `this.APIService.getUserStats` (like those to calls on startup) as you can store `maxListens` on the state once in `componentDidMount`

Reply:
This really boosted the page speed, thanks for the tip :rocket: